library/core/src/iter/range/ptr.rs

```diff
@@ -0,0 +1,274 @@
+use crate::cmp::Ordering;
+use crate::iter::{FusedIterator, TrustedLen, TrustedRandomAccess, TrustedRandomAccessNoCoerce};
+use crate::mem;
+use crate::ops::Range;
+
+macro_rules! impl_iterator_for_ptr_range {
+    ($mutability:ident /* const or mut */) => {
+        /// Iteration of a pointer range, as is common in code that interfaces
+        /// with C++ iterators.
+        ///
+        /// # Safety
+        ///
+        /// Traversing a pointer range is always safe, but **using the resulting
+        /// pointers** is not!
+        ///
+        /// The pointers between the start and end of a range "remember" the
+        /// [allocated object] that they refer into. Pointers resulting from
+        /// pointer arithmetic must not be used to read or write to any other
+        /// allocated object.
+        ///
+        /// As a consequence, pointers from a range traversal are only
+        /// dereferenceable if start and end of the original range both point
+        /// into the same allocated object. Dereferencing a pointer obtained via
+        /// iteration when this is not the case is Undefined Behavior.
```

Comment:
I believe the documented requirement is not excessively restrictive because all the practical use cases probably won't be having `start` and `end` from different allocated objects. If someone does have that situation we could consider relaxing this. For now I thought I would leave open the possibility of implementing `next_back` later in a cleverer way than what I've thought of, instead of division and wrapping_add from the start.

Added an explanation of the alignment to the documentation in 1f15e7332c0a21c6b2beb7c3d9b7031e54d36d6d.